<a href="https://colab.research.google.com/github/Shambhuraje1919/Machine_Learning_Paractice/blob/main/Study_pred_5Models.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import kagglehub

# Download latest version
path = kagglehub.dataset_download("shambhurajejagadale/student-performance-prediction-dataset")

print("Path to dataset files:", path)

Using Colab cache for faster access to the 'student-performance-prediction-dataset' dataset.
Path to dataset files: /kaggle/input/student-performance-prediction-dataset


In [61]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from xgboost import XGBClassifier
from sklearn.metrics import confusion_matrix, confusion_matrix , classification_report

In [4]:
import os

In [5]:
os.listdir(path)

['student_dataset_10000_rows.csv']

In [8]:
df = pd.read_csv(os.path.join(path,"student_dataset_10000_rows.csv"))

In [9]:
df.head(5)

,study_hours,attendance,sleep_hours,internet_usage,assignments_completed,previous_score,exam_score,placement_status
0,7,56,8,7,10,62,100.00,Placed
1,4,69,5,3,8,56,100.00,Placed
2,11,60,7,6,10,45,100.00,Placed
3,8,99,9,8,4,55,90.17,Placed
4,5,52,8,6,8,40,78.82,Placed


In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 8 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   study_hours            10000 non-null  int64  
 1   attendance             10000 non-null  int64  
 2   sleep_hours            10000 non-null  int64  
 3   internet_usage         10000 non-null  int64  
 4   assignments_completed  10000 non-null  int64  
 5   previous_score         10000 non-null  int64  
 6   exam_score             10000 non-null  float64
 7   placement_status       10000 non-null  object 
dtypes: float64(1), int64(6), object(1)
memory usage: 625.1+ KB


In [17]:
df.columns

Index(['study_hours', 'attendance', 'sleep_hours', 'internet_usage',
       'assignments_completed', 'previous_score', 'exam_score',
       'placement_status'],
      dtype='object')

In [92]:
x = df.drop("placement_status_Placed",axis=1)
x_new = df.drop(columns=["placement_status_Placed","exam_score"],axis=1)

This is synthetic dataset thats why its focusing only on exa, score , 70 plus score means placed thats why i have dropped that dataset

In [93]:
x_new.head(2)

,study_hours,attendance,sleep_hours,internet_usage,assignments_completed,previous_score
0,7,56,8,7,10,62
1,4,69,5,3,8,56


In [49]:
df = pd.get_dummies(df,"placement_status",drop_first=True)

In [94]:
df.columns

Index(['study_hours', 'attendance', 'sleep_hours', 'internet_usage',
       'assignments_completed', 'previous_score', 'exam_score',
       'placement_status_Placed'],
      dtype='object')

In [75]:
y = df["placement_status_Placed"]

In [95]:
x_train,x_test,y_train , y_test = train_test_split(x_new,y,test_size = 0.2,random_state=42)

In [96]:
models = {
"model1" : LogisticRegression(),
"model2" : DecisionTreeClassifier(),
"model3" : RandomForestClassifier(),
"model4" : KNeighborsClassifier(),
"model5" : XGBClassifier(),
}


In [97]:
for name ,model in models.items():
  model.fit(x_train,y_train)
  train_score =model.score(x_train ,y_train)
  test_score = model.score(x_test,y_test)
  print(model)
  print("Train Accuracy:",train_score)
  print("Test Accuracy :", test_score)
  print("total diff",(test_score-train_score))
  y_pred = model.predict(x_test)
  print(confusion_matrix(y_pred,y_test))



LogisticRegression()
Train Accuracy: 0.9085
Test Accuracy : 0.9025
total diff -0.006000000000000005
[[ 210   60]
 [ 135 1595]]
DecisionTreeClassifier()
Train Accuracy: 1.0
Test Accuracy : 0.8655
total diff -0.13449999999999995
[[ 204  128]
 [ 141 1527]]
RandomForestClassifier()
Train Accuracy: 1.0
Test Accuracy : 0.8925
total diff -0.10750000000000004
[[ 192   62]
 [ 153 1593]]
KNeighborsClassifier()
Train Accuracy: 0.920125
Test Accuracy : 0.873
total diff -0.04712499999999997
[[ 173   82]
 [ 172 1573]]
XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, device=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric=None, feature_types=None,
              feature_weights=None, gamma=None, grow_policy=None,
              importance_type=None, interaction_constraints=None,
              learning_rate=None, max_bin=None, max_cat_threshold=None,
        

In [98]:
(y).unique()

array([ True, False])

In [99]:
print(y.value_counts())

placement_status_Placed
True     8356
False    1644
Name: count, dtype: int64


In [100]:
print(confusion_matrix(y_pred,y_test))

[[ 193   73]
 [ 152 1582]]


In [102]:
from sklearn.ensemble import RandomForestClassifier
import pandas as pd

rf = RandomForestClassifier(random_state=42)
rf.fit(x_train, y_train)

importance = pd.DataFrame({
    'Feature': x_new.columns,
    'Importance': rf.feature_importances_
})

print(importance.sort_values('Importance', ascending=False))

                 Feature  Importance
0            study_hours    0.250888
5         previous_score    0.197454
4  assignments_completed    0.194072
1             attendance    0.175685
3         internet_usage    0.102573
2            sleep_hours    0.079328


In [10000]:
print(df.groupby('placement_status_Placed')['study_hours'].describe())

                          count      mean       std  min  25%  50%  75%   max
placement_status_Placed                                                      
False                    1644.0  3.130170  2.117967  1.0  1.0  3.0  4.0  11.0
True                     8356.0  6.552178  3.028219  1.0  4.0  7.0  9.0  11.0


| Model               | Train Acc |   Test Acc | Overfitting |
| ------------------- | --------: | ---------: | ----------: |
| Logistic Regression |    90.85% | **90.25%** |    Very Low |
| Decision Tree       |   100.00% |     86.55% |        High |
| Random Forest       |   100.00% |     89.25% |    Moderate |
| KNN                 |    92.01% |     87.30% |         Low |
| XGBoost             |    98.53% |     88.75% |    Moderate |
